exp/etcdrestore/controllers/etcdmachinesnapshot_controller.go

```diff
@@ -60,18 +68,193 @@ func (r *EtcdMachineSnapshotReconciler) SetupWithManager(_ context.Context, mgr
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(etcdMachineSnapshot, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		// Patch ObservedGeneration only if the reconciliation completed successfully.
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+		}
+		if err := patchHelper.Patch(ctx, etcdMachineSnapshot, patchOpts...); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Handle deleted etcdMachineSnapshot
+	if !etcdMachineSnapshot.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, etcdMachineSnapshot)
+	}
+
+	// Ensure the finalizer is present
+	if !controllerutil.ContainsFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer) {
+		controllerutil.AddFinalizer(etcdMachineSnapshot, snapshotrestorev1.ETCDMachineSnapshotFinalizer)
+		if err := r.Client.Update(ctx, etcdMachineSnapshot); err != nil {
+			log.Error(err, "Failed to add finalizer to EtcdMachineSnapshot")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return r.reconcileNormal(ctx, etcdMachineSnapshot)
 }
 
 func (r *EtcdMachineSnapshotReconciler) reconcileNormal(
-	_ context.Context, _ *snapshotrestorev1.EtcdMachineSnapshot,
-) (_ ctrl.Result, err error) {
+	ctx context.Context, etcdMachineSnapshot *snapshotrestorev1.EtcdMachineSnapshot,
+) (ctrl.Result, error) {
+
+	// Handle different phases of the etcdmachinesnapshot creation process
+	switch etcdMachineSnapshot.Status.Phase {
+	case "":
+		// Initial phase, set to Pending
+		etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhasePending
+	case snapshotrestorev1.ETCDSnapshotPhasePending:
+		// Transition to Running
+		etcdMachineSnapshot.Status.Phase = snapshotrestorev1.ETCDSnapshotPhaseRunning
+	case snapshotrestorev1.ETCDSnapshotPhaseRunning:
+		// Check the status of the snapshot creation process
+		// List ETCDSnapshotFile resources to determine if the snapshot is complete
+		if err := checkSnapshotStatus(ctx, r, etcdMachineSnapshot); err != nil {
+			// If no matching ready snapshot is found, requeue the request
+			return ctrl.Result{RequeueAfter: snapshotPhaseRequeueDuration}, nil
+		}
+		return ctrl.Result{}, nil
+	case snapshotrestorev1.ETCDSnapshotPhaseFailed:
+		// If the snapshot creation failed, requeue the request
+		return ctrl.Result{RequeueAfter: snapshotPhaseRequeueDuration}, nil
+	case snapshotrestorev1.ETCDSnapshotPhaseDone:
+		// Snapshot creation is complete, no further action needed
+		return ctrl.Result{}, nil
+	}
+
+	patchBase := client.MergeFromWithOptions(etcdMachineSnapshot.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+	// Patch the resource at the top-level
+	if err := r.Client.Status().Patch(ctx, etcdMachineSnapshot, patchBase); err != nil {
```

Comment:
```suggestion
	
```

Reply:
no need to patch here if the object will be patched at the top-level Reconcile()